File: my-folder/0623-add-one-row-to-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if (root == null || depth == 1) {
            TreeNode node = new TreeNode(val);
            node.left = root;
            return node;
        } 
        dfs(root, val, depth, 1);
        return root;
    }

    private void dfs(TreeNode root, int val, int depth, int current) {
        if (root == null) {
            return;
        }
        if (depth - current == 1) {
            TreeNode left = root.left;
            TreeNode right = root.right;
            root.left = new TreeNode(val);
            root.right = new TreeNode(val);
            root.left.left = left;
            root.right.right = right;
            return;
        }
        dfs(root.left, val, depth, current+1);
        dfs(root.right, val, depth, current+1);
    }
}

File: my-folder/0623-add-one-row-to-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1:
            node = TreeNode(val)
            node.left = root
            return node
        def dfs(node, level, depth, val):
            if not node:
                return
            if level + 1 == depth:
                left = node.left
                right = node.right
                node.left = TreeNode(val)
                node.right = TreeNode(val)
                node.left.left = left
                node.right.right = right
            dfs(node.left, level + 1, depth, val)
            dfs(node.right, level + 1, depth, val)
        dfs(root, 1, depth, val)
        return root
